core/src/main/java/com/linecorp/armeria/server/HttpHeaderInjectingRpcService.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2018 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server;
+
+import javax.annotation.Nullable;
+
+import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.Request;
+import com.linecorp.armeria.common.Response;
+import com.linecorp.armeria.common.RpcRequest;
+import com.linecorp.armeria.common.RpcResponse;
+
+/**
+ * Decorates a {@link Service} to inject {@link HttpHeaders} when using RPC which uses HTTP as underlying
+ * protocol. The {@link HttpHeaders} returned from
+ * {@link #httpHeaders(ServiceRequestContext, RpcRequest, RpcResponse)} will be added to the underlying
+ * {@link HttpResponse}.
+ * Here is an example which uses {@code ThriftCallService} and {@code THttpService} in armeria-thrift module:
+ *
+ * <pre>{@code
+ * final ServerBuilder sb = ...
+ * sb.service("/", ThriftCallService.of(thriftHandler)
+ *                                  .decorate(...)
+ *                                   // A subclass of HttpHeaderInjectingRpcService
+ *                                  .decorate(MyInjectingService::new)
```

Comment:
A user would not need to create a new class if he or she can build a decorator with a function. (See the comment below)